### DISCUSSION TITLE
Explore pyright for CI and command line type checking in Python

### DISCUSSION THREAD

**Discussion by prrao87** [thedataquarry/rustinpieces#31] [Ideas]
For CI and command line type checking, it makes sense to explore pyright a bit more, especially as they've been [continually updating this page](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md).

There's enough evidence that pyright has significantly improved from its earlier days, and it's also likely to be faster than mypy and it's very actively being maintained and improved.

This particular claim from this [Dagster blog](https://dagster.io/blog/adding-python-types) is one I'd like to test in the future:

> Unless you have an existing mypy workflow or configuration, we recommend skipping mypy entirely and just using pyright (for command line type-checking and a language server for non-VSCode editors) and pylance (as a language server with VSCode). It’s not just the language server functionality— pyright’s type-checking is also better than mypy's. Pyright is faster, has a much more responsive development team (bugs are fixed fast; mypy bugs linger on the issue tracker for years), and understands certain typing constructs (particularly recursive types) better than mypy.

I think those reasons of rapid responses to bugs and handling recursive types is why Pydantic made the choice to base their checks on pyright and not mypy.

**Comment by prrao87** [comment]
Related to #29

**Comment by sanders41** [comment]
These reasons have me intrigued. On a quick look my main concern is if you `pip install pyright` it installs node onto your system https://microsoft.github.io/pyright/#/installation?id=python-package. Installing node when you pip install something seems wrong. 

I have node and pyright installed for my lsp (I don't use VS Code so no pylance) so this isn't a big deal for me. But I could see it being a barrier to entry for contributions on an open source project from Python developers that want nothing to do with node/js.

**Reply by prrao87** [reply]
Ugh, that doesn't seem very appealing at all (node for Python users???). Will take a look at it too, but mypy is good for now!

**Comment by prrao87** [comment]
The supposed speed differences between mypy and pyright would only be noticeable for really large codebases. Also, the recursive type checks and such are useful for projects that require that kind of complexity. I think it's best to keep this on the backburner for now. I've wanted to try out pyright for a while now, but this repo may not be the best place to do it 😅.

Let's keep this open for future reference anyway.
